Deploy da Função Lambda de Listagem de Tarefas

GrazielleNascimento opened the pull request:
# Descrição do PR - **Deploy da Função Lambda de Listagem de Tarefas**

## Objetivo:
Este Pull Request tem como objetivo a implementação, configuração e deploy de uma **função Lambda** para listagem de tarefas a partir de uma tabela DynamoDB. A função será acessada via **API Gateway**, permitindo a listagem das tarefas no formato JSON. Também foi implementada a estrutura necessária para garantir a segurança e autenticação via **AWS Cognito**.

## Alterações Realizadas:

### 1. **Função Lambda - `list_item.py`**:
   - **Criação da Função Lambda** para listar tarefas, que recebe parâmetros de **data** e **user_id** (via query string ou corpo da requisição).
   - A função acessa o **DynamoDB** para buscar as tarefas com base na data ou em todas as tarefas do usuário.
   - O código foi estruturado para retornar a lista de tarefas ou uma mensagem de erro caso ocorra alguma falha na consulta ou no acesso ao banco de dados.
   - **Fluxo da função `listar_tarefas`**:
     - Se um **parâmetro de data** for fornecido, a função faz uma consulta (`query`) no DynamoDB utilizando a chave primária e o filtro de itens (usando `SK` que começa com `ITEM#`).
     - Caso contrário, se nenhum parâmetro de data for fornecido, a função faz um **scan** em toda a tabela.

### 2. **API Gateway e Lambda Integration**:
   - O API Gateway foi configurado para fornecer acesso público a essa Lambda, onde será possível consumir as tarefas via **requisição GET** no endpoint `/lista-tarefa`.
   - A API é protegida com autenticação **Cognito**, o que requer a obtenção de um **token JWT** para acessar os dados.

### 3. **Testes**:
   Foram realizados **19 testes unitários e de integração** para garantir que a função Lambda e o API Gateway estão funcionando corretamente.

   **Testes de Sucesso:**
   - **Testar Listagem de Tarefas por Data**:
     - **Teste 1**: Listar tarefas para a data específica `2024-12-26`. O resultado foi de 2 tarefas com sucesso.
   - **Testar Listagem de Tarefas sem Data**:
     - **Teste 2**: Listar todas as tarefas, retornando 3 tarefas com sucesso.
   - **Testar Lista Vazia**:
     - **Teste 3**: Listar tarefas para uma data sem tarefas (`2024-12-25`), retornando uma lista vazia.
   - **Testar Listagem com `user_id`**:
     - **Teste 4**: Listar tarefas com o `user_id`, retornando todas as 3 tarefas.
   - **Testar Formatação da Resposta**:
     - **Teste 5**: Validar a estrutura da resposta, garantindo que os campos obrigatórios estão presentes e formatados corretamente.
   - **Testar Conversão de Data para PK**:
     - **Teste 6**: Validar se a conversão de data para o formato correto de chave primária (`PK`) está funcionando corretamente.

   **Testes de Falha:**
   - **Teste de Erro ao Consultar o DynamoDB (Query)**:
     - **Teste 7**: Simulação de erro ao consultar o DynamoDB com uma exceção de conexão.
   - **Teste de Erro ao Consultar o DynamoDB (Scan)**:
     - **Teste 8**: Simulação de erro durante o scan no DynamoDB, retornando "Tabela não encontrada".
   - **Teste de Erro ao Acessar a Tabela**:
     - **Teste 9**: Simulação de erro de permissão ao acessar a tabela no DynamoDB.

   **Testes de Lambda Handler:**
   - **Testar com Parâmetros de Query**:
     - **Teste 10**: Testar o Lambda com parâmetros de query, simulando a chamada via API Gateway com `data` na query string.
   - **Testar com Parâmetros no Corpo**:
     - **Teste 11**: Testar o Lambda com parâmetros no corpo da requisição.
   - **Testar com Usuário Cognito**:
     - **Teste 12**: Testar o Lambda com autenticação Cognito, verificando a extração do `user_id` do JWT.
   - **Testar com Cabeçalhos CORS**:
     - **Teste 13**: Validar se os cabeçalhos CORS estão presentes e configurados corretamente.
   - **Testar com Content-Type Correto**:
     - **Teste 14**: Validar se o `Content-Type` está correto na resposta da Lambda.

   **Testes de Erro no Lambda Handler:**
   - **Testar Erro Interno no Lambda**:
     - **Teste 15**: Simulação de erro interno na função Lambda, retornando status 500.
   - **Testar JSON Inválido no Corpo**:
     - **Teste 16**: Simulação de erro ao passar um JSON inválido no corpo da requisição.
   - **Testar Resposta com Cabeçalhos CORS**:
     - **Teste 17**: Verificação de que os cabeçalhos CORS estão presentes e corretamente configurados.
   - **Testar Parâmetros de Query no Lambda Handler**:
     - **Teste 18**: Testar o Lambda com parâmetros de query string, verificando a extração e uso dos mesmos.
   - **Testar Sem Parâmetros no Lambda Handler**:
     - **Teste 19**: Testar o Lambda sem parâmetros, retornando todas as tarefas.

![image](https://github.com/user-attachments/assets/5eda37f2-386d-4533-89ac-c119c2517386)


### 4. **Terraform**:
   - **Infraestrutura configurada no Terraform**:
     - Módulo de criação da função Lambda.
     - Módulo de criação do **API Gateway** com integração à função Lambda.
     - Módulo de autenticação via **AWS Cognito** para garantir a segurança da API.
     - O **role IAM** necessário para que a função Lambda tenha permissão para acessar o DynamoDB e o API Gateway foi configurado.
   - **Saídas do Terraform**:
     - A URL do endpoint da API Gateway para listar as tarefas foi definida como saída do Terraform, para facilitar o acesso durante os testes.

### 5. **Configuração do Ambiente**:
   - Variáveis de ambiente como `TABLE_NAME` (nome da tabela do DynamoDB) e `AWS_DEFAULT_REGION` (região da AWS) foram configuradas para garantir a execução correta da função Lambda.
   - Certifiquei-me de que o ambiente de execução da função Lambda foi configurado corretamente com a versão do Python (`python3.12`) e com a quantidade de memória e tempo de execução adequados.

### 6. **Scripts de Debug Local**:
   - Foi criada uma versão de **debug local** da função Lambda, permitindo que o desenvolvedor consiga testar a função Lambda localmente sem a necessidade de credenciais ou deploy na AWS.
   - Utilizando **mocks**, os testes são executados com dados simulados, permitindo verificar se a lógica está funcionando corretamente sem depender da infraestrutura real.
   
   
![image](https://github.com/user-attachments/assets/a365d49d-e93f-437c-bff0-ba11d9750e97)


### 7. **Pipeline de CI/CD (GitHub Actions)**:
   - A pipeline foi configurada para ser acionada **automaticamente** sempre que um PR for aberto ou atualizado na branch `dev`.
   - A pipeline executa os seguintes passos:
     - Instalação e configuração do Terraform.
     - Execução do `terraform init`, `terraform validate`, `terraform plan` e `terraform apply`.
     - Deploy da função Lambda, do API Gateway e dos recursos associados.

## Como Testar:

1. **Subir o PR**: Após abrir o PR, a pipeline será acionada automaticamente para realizar o deploy dos recursos na AWS.
2. **Testar no Postman**:
   - **Obter a URL do endpoint** do API Gateway (disponível nos outputs do Terraform).
   - **Configurar o Postman** para fazer uma requisição GET para o endpoint `/lista-tarefa`.
   - **Autenticação**: Utilize um **token JWT** do Cognito (gerado com o usuário configurado) para acessar o endpoint, caso a autenticação esteja ativada.
3. **Verificar a Função Lambda**:
   - No console da AWS, verifique se a função Lambda foi criada corretamente.
   - Verifique os **logs no CloudWatch** para mais detalhes sobre a execução da função Lambda.

## Resultado Esperado:
- Ao realizar uma requisição GET no endpoint `/lista-tarefa`, você deverá ver a lista de tarefas conforme esperado:
  - **Status 200** com as tarefas listadas.
  - Caso não existam tarefas para a data fornecida, o retorno será **status 200** com `tarefas: []`.
  - Caso ocorra algum erro (exemplo, erro no DynamoDB), o status será **500** com a mensagem de erro detalhada.

---

## Tarefas Pendentes: 
- **Verificação Manual**: Após o deploy, a responsável pelo repositório @gaby-lorraynne precisará testar a funcionalidade diretamente na AWS, consumindo o endpoint via Postman ou outra ferramenta.

### Ordem de Execução no Postman:

1. 🔐 **POST Login - Obter Token (PRIMEIRO!)**
2. 📝 **GET /Listar Todas as Tarefas**
3. 📅 **GET /Listar Tarefas por Data**
4. **Testes de Erro**

### Configurar variáveis no Postman:

Após executar o script de deploy:

1. **Obter Informações da API**:
    ```bash
    cd terraform  # ou meu-projeto/terraform
    terraform output
    ```

- `base_url`: URL do output do terraform
- `cognito_client_id`: Client ID do output
- `username`: Email que você criou
- `password`: Senha que você definiu

### 🔐 PASSO 1: Obter Token

**Request 1: Login**

- **Método**: POST
- **URL**: `https://cognito-idp.sa-east-1.amazonaws.com/`

**Headers**:
- `Content-Type`: `application/x-amz-json-1.1`
- `X-Amz-Target`: `AWSCognitoIdentityProviderService.InitiateAuth`

**Body (JSON)**:
    ```json
    {
        "AuthFlow": "USER_PASSWORD_AUTH",
        "ClientId": "SEU_CLIENT_ID",
        "AuthParameters": {
            "USERNAME": "seu.email@exemplo.com",
            "PASSWORD": "SuaSenha123!"
        }
    }
    ```

➡️ **Copiar o AccessToken da resposta**

### 📝 PASSO 2: Testar Endpoint

**Request 2: Lista Tarefas**

- **Método**: GET
- **URL**: `https://SUA-API-ID.execute-api.sa-east-1.amazonaws.com/dev/lista-tarefa`
- **Authorization**: `Bearer SEU_ACCESS_TOKEN`

**Resposta Esperada**:
    ```json
    {
        "message": "Tarefas listadas com sucesso!",
        "count": 0,
        "tarefas": []
    }
    ```

### 📅 PASSO 3: Testar com Data

**Request 3: Lista por Data**

- **Método**: GET
- **URL**: `https://SUA-API-ID.execute-api.sa-east-1.amazonaws.com/dev/lista-tarefa?data=2024-12-26`
- **Authorization**: `Bearer SEU_ACCESS_TOKEN`

GrazielleNascimento committed:
docs: update info with how use pytest

GrazielleNascimento committed:
fix: add missing newline at end of .gitignore file

GrazielleNascimento committed:
fix: add missing newline at end of test_list_tasks.py

GrazielleNascimento committed:
fix: add missing newline at end of list_item.py

GrazielleNascimento committed:
feat: add debug script for local Lambda functions with example file creation

GrazielleNascimento committed:
refactor: reorganize imports and clean up code formatting in list_item.py

GrazielleNascimento committed:
feat: add Lambda function for listing items with archive configuration

GrazielleNascimento committed:
feat: add Lambda and API Gateway modules for listing items

GrazielleNascimento committed:
feat: add outputs for Lambda list item function and API Gateway endpoint

GrazielleNascimento committed:
refactor: reorganize imports and improve code formatting across multiple files

GrazielleNascimento committed:
feat: add outputs for list item Lambda function

GrazielleNascimento committed:
feat: add table_name variable for DynamoDB configuration

GrazielleNascimento committed:
refactor: standardize dictionary key formatting and improve import organization in test_list_tasks.py

GrazielleNascimento committed:
refactor: reorganize imports and standardize string formatting in add_item.py

GrazielleNascimento committed:
refactor: reorganize imports and improve JSON response formatting in delete_item.py

GrazielleNascimento committed:
refactor: streamline body and response formatting in hello_terraform.py

GrazielleNascimento committed:
refactor: reorganize imports and improve formatting in update_item.py

GrazielleNascimento committed:
docs: update testing and utility commands in README.md for clarity and organization

GrazielleNascimento committed:
fix: standardize environment variable naming for TABLE_NAME in lambda_list_item.tf

GrazielleNascimento commented on meu-projeto/lambda/lambda_delete_item/delete_item.py:
```diff
@@ -1,22 +1,26 @@
+import boto3
+
+dynamodb = boto3.client("dynamodb", region_name="sa-east-1")
+TABLE_NAME = os.environ.get("TABELA_LISTA", "ListaMercado")
```
@gaby-lorraynne  oq acha?  vai impactar em algo do seu repo?

gaby-lorraynne reviewed (approved):
Aprovado!

GrazielleNascimento committed:
feat: add IAM role policy attachment for DynamoDB access in main.tf

GrazielleNascimento committed:
feat: add Lambda function and archive for listing items in main.tf

GrazielleNascimento committed:
fix: rename IAM role policy attachment for DynamoDB access for consistency

GrazielleNascimento committed:
refactor: remove unused variables for role ARN and DynamoDB table name in variables.tf

GrazielleNascimento committed:
refactor: remove Lambda function and related archive for listing items in lambda_list_item.tf

GrazielleNascimento committed:
feat: add conditional creation for listar_item Lambda function and update outputs

GrazielleNascimento committed:
refactor: remove redundant ZIP data block for list_item Lambda function and add table_name variable to hello_terraform module

GrazielleNascimento committed:
fix: implementation of api gateway

GrazielleNascimento committed:
refactor: remove dependency on hello_terraform Lambda function in API Gateway permission

GrazielleNascimento committed:
refactor: simplify dependency for API Gateway deployment

GrazielleNascimento committed:
chore: add a blank line for better readability before lambda permission resource

GrazielleNascimento committed:
feat: add Lambda function for listing items with zip packaging

GrazielleNascimento committed:
refactor: remove unused Lambda function and outputs for item listing

GrazielleNascimento committed:
refactor: remove unused role_arn variable from Lambda configuration

GrazielleNascimento committed:
refactor: consolidate Lambda output variables and remove unused item listing outputs

GrazielleNascimento committed:
refactor: remove unused Lambda output variables

GrazielleNascimento committed:
refactor: update Lambda module output variable names for consistency

GrazielleNascimento committed:
refactor: update output variable names for Lambda list item module for consistency

GrazielleNascimento committed:
refactor: update Lambda list item module output variable names for consistency

GrazielleNascimento committed:
refactor: create IAM role for Lambda function and update role ARN reference in main module

GrazielleNascimento committed:
refactor: update IAM role references for Lambda function to ensure consistency

GrazielleNascimento committed:
refactor: update Lambda function name to include version for clarity

GrazielleNascimento committed:
refactor: enhance query logic in listar_tarefas to filter tasks by date

GrazielleNascimento committed:
refactor: improve query logic in listar_tarefas to ensure correct date filtering

GrazielleNascimento committed:
refactor: enhance listar_tarefas to improve task retrieval and filtering logic

GrazielleNascimento committed:
refactor: optimize listar_tarefas to improve task filtering by adding FilterExpression for date

GrazielleNascimento committed:
refactor: enhance listar_tarefas to improve data filtering and validation logic

JefersonKlaus commented on meu-projeto/lambda/lambda_list_item/list_item.py:
```diff
@@ -26,17 +26,21 @@ def listar_tarefas(data=None, user_id=None):
-                FilterExpression=Key("PK").begins_with("LIST#")
+                 FilterExpression=Attr("PK").begins_with("LIST#") & Attr("SK").begins_with("ITEM#")
             )
+
```
SCAN em dynamo é extremamente ineficiente, esta ciente disso?

GrazielleNascimento committed:
refactor: streamline listar_tarefas by optimizing query and scan logic, removing debug logs

GrazielleNascimento committed:
refactor: enhance listar_tarefas to support multiple date formats and improve data filtering

GrazielleNascimento committed:
refactor: enhance listar_tarefas with improved logging and data filtering for date queries

GrazielleNascimento committed:
feat: add API Gateway module for listing items with Cognito authorization

GrazielleNascimento committed:
feat: add function_name_list variable and update API Gateway module to use it

GrazielleNascimento committed:
fix: update lista_tarefa_endpoint to use correct API invoke URL

GrazielleNascimento committed:
refactor: rename function_name variable to function_name_list for consistency and update default value

GrazielleNascimento committed:
fix: update DynamoDB region from sa-east- across multiple Lambda functions and Terraform configurations
refactor: enhance Lambda function roles and policies with unique names and improved variable descriptions
feat: add environment variable support for table names and improve resource definitions in Terraform
delete: remove unused lambda_list_item module and related configurations
